tikv/raftstore/fsm.go

```diff
@@ -0,0 +1,556 @@
+	if mailbox.len() == checkedLen {
+		return true
+	}
+	fsm = mailbox.takeFsm()
+	if fsm == nil {
+		return true
+	}
+	fsm.setMailbox(mailbox)
+	b.swapInsert(index, fsm)
+	return false
+}
+
+func (b *batch) swapRemove(index int) fsm {
+	fsm := b.normals[index]
+	lastIdx := len(b.normals) - 1
+	b.normals[index] = b.normals[lastIdx]
+	b.normals = b.normals[:lastIdx]
+	return fsm
+}
+
+func (b *batch) swapInsert(index int, fsm fsm) {
+	lastIdx := len(b.normals)
+	b.normals = append(b.normals, fsm)
+	b.normals[index], b.normals[lastIdx] = b.normals[lastIdx], b.normals[index]
+}
+
+/// Remove the normal FSM located at `index`.
+///
+/// This method should only be called when the FSM is stopped.
+/// If there are still messages in channel, the FSM is untouched and
+/// the function will return false to let caller to keep polling.
+func (b *batch) remove(index int) bool {
+	fsm := b.swapRemove(index)
+	mailbox := fsm.takeMailbox()
+	if mailbox.isEmpty() {
+		mailbox.release(fsm)
+		return true
+	}
+	fsm.setMailbox(mailbox)
+	b.swapInsert(index, fsm)
+	return false
+}
+
+func (b *batch) releaseControl(controlBox *mailbox, checkedLen int) bool {
+	fsm := b.control
+	b.control = nil
+	controlBox.release(fsm)
+	if controlBox.len() == checkedLen {
+		return true
+	}
+	fsm = controlBox.takeFsm()
+	if fsm == nil {
+		return true
+	}
+	b.control = fsm
+	return false
+}
+
+func (b *batch) removeControl(controlBox *mailbox) {
+	if controlBox.isEmpty() {
+		fsm := b.control
+		b.control = nil
+		controlBox.release(fsm)
+	}
+}
+
+/// A handler that poll all FSM in ready.
+///
+/// A General process works like following:
+/// ```text
+/// loop {
+///     begin
+///     if control is ready:
+///         handle_control
+///     foreach ready normal:
+///         handle_normal
+///     end
+/// }
+/// ```
+///
+/// Note that, every poll thread has its own handler, which doesn't have to be
+/// Sync.
+type pollHandler interface {
+
+	/// This function is called at the very beginning of every round.
+	begin(batchSize int)
+
+	/// This function is called when handling readiness for control FSM.
+	///
+	/// If the returned value pause is true, the chLen represents the current channel length.
+	/// This function will only be called for the same fsm after channel's length is larger than chLen.
+	/// If pause is false, then this function will still be called for the same FSM in the next
+	/// loop unless the FSM is stopped.
+	handleControl(control fsm) (pause bool, chLen int)
+
+	/// This function is called when handling readiness for normal FSM.
+	///
+	/// The returned value is handled in the same way as `handle_control`.
+	handleNormal(normal fsm) (pause bool, chLen int)
+
+	/// This function is called at the end of every round.
+	end(normals []fsm)
+}
+
+type poller struct {
+	router       *router
+	fsmReceiver  <-chan *Msg
+	handler      pollHandler
+	maxBatchSize int
+}
+
+func (p *poller) fetchBatch(batch *batch, maxSize int) {
+	msg, ok := <-p.fsmReceiver
```

Comment:
This rely on an assumption that there must be some messages in the channel. The latest TiKV has some changes for this function, please sync it.